### DISCUSSION TITLE
ECMAScript.2015

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#3] [프론트 기술 질문 공간]
es6의 변경점에는 어떤 것이 있는지 알려주시고 해당 내용에 대해 설명해주세요

**Comment by javascriptbiggosoo** [comment]
이제는 es6 출시된지 오랜시간이 지났지만 여전히 면접에서 많이 묻는 질문이라고 알고있습니다.
대신 이제는 사용하지 못하는 사람이 더 적어졌다고 할 수 있을만큼 묻는 의도는 바뀌었다고 보는데요, 문법의 숙지여부 보다는 '왜 등장했는가?'를 설명을 할 수 있느냐가 중요시 될 것 같습니다.

"ECMA Script는 꾸준히 개선이 이루어지고 있는데, 그중 ES6 버전은 최신 문법을 대표하는 말로 자리 잡았을 만큼 많은 변화를 가져왔습니다.
이전까지 웹에서 쭉 사용된 언어의 작동 방식을 건드릴 수는 없으니 생성자함수를 대체하는 class, promise의 then catch 핸들링을 async/await로 대체한 것과 같은 문법적 설탕을 추가했으며 _array_를 편하게 다룰 수 있도록 해주는 map, filter, reduce메소드의 추가, _object_ 리터럴의 보강 등 자주 사용하는 여러 기능에 더욱 편의성을 증가시켜주었습니다.
이외에도 수많은 변화가 있었는데 그 배경에는 전에비해 프론트엔드 개발의 중요성이 커졌기 때문에 언어의 활용성을 증가시켰다는 점을, 이로인한 효과로는 다른 언어를 사용하던 개발자들의 접근성을 증가시켰다는 점을 꼽을 수 있습니다.
여러 언어들이 수렴 진화를 하고 있는 와중 TS와 ES6가 등장하면서 JS는 단연 가장 큰 변화가 일어난 언어가 되었다고 말할 수 있습니다."


정도로 답을 할 것 같습니다.
추가로 이 답변이 괜찮은지, 보강한다면 어떨지에 해대 AskUp(GPT3.5)에 물어본 결과도 공유합니다. ㅎㅎ

"이 답변은 대체로 괜찮은 내용을 담고 있지만, 더 자세하게 설명할 수 있는 부분이 있습니다.

첫째로, ES6에서 추가된 기능 중에서도 중요한 것들을 꼽아서 설명하는 것이 좋습니다. 예를 들어, let/const 변수 선언 방식, 화살표 함수, 클래스, 프로미스, 디스트럭처링, rest/spread 연산자 등이 있습니다.

둘째로, ES6 이외에도 ES7, ES8 등 이후 버전에서 추가된 기능들도 언급하면 좋습니다. 예를 들어, async/await, 객체의 entries/values 메서드, 지수 연산자, 객체 rest/spread 연산자 등이 있습니다.

셋째로, ES6에서 추가된 기능들이 기존의 언어보다 어떤 장점을 가지고 있는지 구체적으로 설명하는 것이 좋습니다. 예를 들어, 클래스를 사용하면 객체 지향 프로그래밍을 더욱 쉽게 할 수 있고, 프로미스를 사용하면 비동기 코드를 더욱 간결하게 작성할 수 있다는 등의 내용을 언급할 수 있습니다.

마지막으로, ES6 이후에도 계속해서 새로운 기능들이 추가되고 있다는 점을 언급하면 좋습니다. ES6는 대표적인 문법 개선 버전이지만, ES7, ES8 등에서도 새로운 기능들이 추가되고 있기 때문에, 개발자는 항상 최신 버전의 언어를 익히는 것이 중요하다는 것을 강조할 수 있습니다."

**Comment by michoball** [comment]
많은 변경사항들이 있지만 개발을 하며 많이 사용하고 있는 것들을 예로 들어보자면

const, let 
-> 변수를 전역 스코프에 저장하여 불필요하게 메모리를 사용하고 중복선언되어 디버깅을 어렵게하 var의 단점을 보완하고자 나온 것

promise, async & await
 -> JS의 비동기 처리를 위해 원래 사용하던 callback 함수의 단점을 보완하고자 나온 문법

arrow 함수
-> 함수를 선언이 아닌 표현문으로 함수를 간편히 작성할 수 있게 해준다.

 그외에 Object spread / rest 등이 있지만 요즘 제일 관심이 있는 것은 배열의 reduce 메서드입니다.  

reduce
-> 배열을 돌며 값을 누적하여 계산해주는 누산기 같은 역할 

단순 값 계산 뿐아니라 배열을 돌며 새로운 배열이나 객체를 만들거나 특정 값을 추출하는 등 많은 응용을 할 수 있게 해주기에 계속 학습하고 다양하게 프로젝트에 적용하려 노력중인 메서드입니다.

---
한두가지 생각나는 것들을 말하고 그것을 설명하는 상황을 생각했는데 막상 적으려니 너무 많네요 ㅎㅎㅎ

**Comment by skyhanull** [comment]
es5와 es6의 차이점이라고 한다면
1. let,const의 등장 => var가 아닌 let과 const의 등장으로 var의 재할당/재선언성을 보완하고자 탄생
2. Arrow Functions의 사용 => Arrow Functions은 자신의 this, arguments을 바인딩하지 않습니다
3. 기존의 반복문에서 for..in/for..of문과 forEach문이 등장했습니다
4. 구조분해할당과 spread연산자가 등장했습니다 
5. 템플릿 리터럴의 등장
6. 비교문이 기존에는 '=='였지만 es6로 오면서 '==='로 바뀌게 되어 값만 아니고 자료형도 비교하게 되었습니다
7. promise , async/await의 등장 특히 async/await의 등장으로 비동기를 동기문처럼 쓸수 있게 되었습니다

**Comment by BaekYuri** [comment]
1. let, const : 재선언 불가, 재할당 가능한 let, 상수 키워드 const 추가
2. 백틱 : `${str}` 형태로 문자열 가독성 증가
3. Arrow function의 추가: 함수 선언 시 가독성 증가
4. 구조 분해 할당 : const {name, id} = users 형태로 객체의 속성을 간결한 방식으로 가져올 수 있음
5. spread operator(...) : 배열을 이어 붙이거나 복사 등을 할 때 간편하게 사용
6. promise : 비동기 처리 시 콜백 지옥을 해소하기 위해 프로미스 문법 등장

ps. async/await는 ES8 문법에서 나온 것으로 알고 있습니다.
